tests/src/Integration/Engine/PrepareMetadataTest.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\Tests\rules\Integration\Engine\PrepareMetadataTest.
+ */
+
+namespace Drupal\Tests\rules\Integration\Engine;
+
+use Drupal\rules\Context\ContextConfig;
+use Drupal\rules\Context\ContextDefinition;
+use Drupal\rules\Engine\ExecutionMetadataState;
+use Drupal\rules\Engine\RulesComponent;
+use Drupal\Tests\rules\Integration\RulesEntityIntegrationTestBase;
+
+/**
+ * Tests that the setup of the execution metadata state for an expression works.
+ *
+ * @group rules
+ */
+class PrepareMetadataTest extends RulesEntityIntegrationTestBase {
+
+  /**
+   * Tests that a variable can be added by an action and is then available.
+   */
+  public function testAddingVariable() {
+    $rule = $this->rulesExpressionManager->createRule();
+    $rule->addAction('rules_variable_add', ContextConfig::create()
+      ->setValue('type', 'string')
+      ->setValue('value', '')
+      ->provideAs('variable_added', 'result')
+    );
+
+    $state = ExecutionMetadataState::create();
+    $found = $rule->prepareExecutionMetadataState($state);
+    $this->assertTrue($state->hasDataDefinition('result'));
+    $this->assertTrue($found);
+  }
+
+  /**
+   * Tests partial state setup until an expression is reached in the tree.
+   */
+  public function testPreparingUntil() {
+    // Setup a rule with 2 actions.
+    $rule = $this->rulesExpressionManager->createRule();
+    $rule->addAction('rules_variable_add', ContextConfig::create()
+      ->setValue('type', 'string')
+      ->setValue('value', '')
+      ->provideAs('variable_added', 'result1')
+    );
+    $second_action = $this->rulesExpressionManager->createAction('rules_variable_add')
+      ->setConfiguration(ContextConfig::create()
```

Comment:
no, end() does not work with iterators. It would have to be "$rule->getIterator()->offsetGet(1);" which is also not very nice. For conditions in a rule "$rule->getConditions()->getIterator()->offsetGet(1);".

I'm leaving this as is for now.